716 rework dfns section (#942)
Dyalog/documentation: programming-reference-guide/docs/defined-functions-and-operators/dfns-and-dops/restrictions.md

- [`⎕SHADOW`](../../../../language-reference-guide/system-functions/shadow) skips dfns when looking down the stack for a tradfn in which to make a new local name.
- Control structures and other `:`-keywords, such as `:If` and `:Return`, are unavailable; a dfn expresses conditions through guards instead.
- Monadic branch (`→` given a line number, equivalent to `:GoTo`) is not supported. Niladic branch, that is [abort](../../../../language-reference-guide/other-syntax/abort) (`→`), does work: it clears the most recently suspended statement and all of its pendent statements from the state indicator.
- Modified assignment (`X f←Y`) and use of the pass-through value of an assignment (as in `X f Y←Z`) behave as expected when `f` is a primitive, but in a dfn a named function is instead read as part of a multiple assignment, so `X plus←10` assigns `10` to both `X` and `plus`. Insert `∘⊢` to force the intended reading: `X plus∘⊢←10` for the modified assignment, and `1 plus∘⊢a←10` (or simply `1 plus⊢a←10`) for the pass-through.

> ## Supplied Workspaces

Reviewer: I see that this Supplied Workspaces section isn't new, but it looks a bit odd to me that it is here, under the restrictions. Or is it because it is meant as a list of examples?

# Restrictions

> A dfn or dop is subject to the following restrictions.

Reviewer 2: ```suggestion
The following restrictions apply to dfns and/or dops:
```

> - A dfn need not return a [result](../../introduction/results.md), but any expression that does not produce a result still terminates the dfn. You therefore cannot call a no-result function from the middle of a dfn and carry on, because evaluation stops at that expression.

Reviewer 2: ```suggestion
- A dfn does not need to return a [result](../../introduction/results.md). However, any expression that does not produce a result still terminates the dfn. This means that you cannot call a no-result function from the middle of a dfn and expect to continue after the dfn has been evaluated, as evaluation stops at that expression.
```

> - Normal tracing (`<TC>`) steps over a single-line dfn, executing it atomically like an [Execute](../../../../language-reference-guide/primitive-functions/execute) (`⍎`) expression. This deliberate restriction avoids the confusion of tracing a line and seeing nothing change. Inline Tracing (`<IT>`) can step through the functions called within a single-line dfn, and a multi-line dfn traces normally.

Reviewer 2: ```suggestion
- Normal tracing (command code **<TC>**) steps over a single-line dfn, executing it atomically like an _execute_ ([`⍎`](../../../../language-reference-guide/primitive-functions/execute)) expression. This deliberate restriction avoids the confusion of tracing a line and seeing nothing change. Inline tracing (command code **<IT>**) can step through the functions called within a single-line dfn, and a multi-line dfn traces normally.
```

> - [`⎕CS`](../../../../language-reference-guide/system-functions/cs) is not supported inside a dfn and signals a `NONCE ERROR`.

Reviewer 2: ```suggestion
- [`⎕CS`](../../../../language-reference-guide/system-functions/cs) is not supported inside a dfn; attempting to call it generates a `NONCE ERROR`.
```

- [`⎕MONITOR`](../../../../language-reference-guide/system-functions/set-monitor), [`⎕TRACE`](../../../../language-reference-guide/system-functions/set-trace), and [`⎕LOCK`](../../../../language-reference-guide/system-functions/lock) do not apply to dfns or dops.
- [`⎕SHADOW`](../../../../language-reference-guide/system-functions/shadow) skips dfns when looking down the stack for a tradfn in which to make a new local name.

> - Control structures and other `:`-keywords, such as `:If` and `:Return`, are unavailable; a dfn expresses conditions through guards instead.

Reviewer 2: ```suggestion
- [Control structures](../traditional-functions-and-operators/control-structures/control-structures-summary/) and other `:`-keywords, such as `:If` and `:Return`, cannot be used within a dfn. Instead, a dfn expresses conditions through [guards](guards.md).
```

- Monadic branch (`→` given a line number, equivalent to `:GoTo`) is not supported. Niladic branch, that is [abort](../../../../language-reference-guide/other-syntax/abort) (`→`), does work: it clears the most recently suspended statement and all of its pendent statements from the state indicator.

> - Modified assignment (`X f←Y`) and use of the pass-through value of an assignment (as in `X f Y←Z`) behave as expected when `f` is a primitive, but in a dfn a named function is instead read as part of a multiple assignment, so `X plus←10` assigns `10` to both `X` and `plus`. Insert `∘⊢` to force the intended reading: `X plus∘⊢←10` for the modified assignment, and `1 plus∘⊢a←10` (or simply `1 plus⊢a←10`) for the pass-through.

Reviewer 2: ```suggestion
- Modified assignment (`X f←Y`) and use of the pass-through value of an assignment (that is, `X f Y←Z`) behave as expected when `f` is a primitive. However, in a dfn a named function is instead read as part of a multiple assignment, so `X plus←10` assigns `10` to both `X` and `plus`. Instead, modified assignment can be achieved by inserting `∘⊢` , that is, by using `X plus∘⊢←10`, and pass-through can be achieve by inserting with `1 plus∘⊢a←10` or `1 plus⊢a←10`.
```
